src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,44 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbOptional = userService.findByEmail(email);
+        if (userFromDbOptional.isEmpty()) {
+            throw new AuthenticationException("Can't authenticate user");
+        }
+        User userFromDB = userFromDbOptional.get();
+        String hashedPassword = HashUtil.getHashPassword(password, userFromDB.getSalt());
+        if (!hashedPassword.equals(userFromDB.getPassword())) {
+            throw new AuthenticationException("Can't authenticate user");
+        }
+        return userFromDB;

```

Comment:
Try to optimize this logic, for example, like this
```suggestion
        if (userFromDbOptional.isPresent() && matchPasswords(email, userFromDbOptional.get())) {
            return userFromDbOptional.get();
        }
        throw new AuthenticationException("Can't authenticate user by email: " + email);

```